app/api/db/migrations/run.py

```diff
@@ -0,0 +1,89 @@
+# Convenience script for running alembic migration commands through a pyscript
+# rather than the command line. This allows poetry to package and alias it for
+# running on the production docker image from any directory.
+import itertools
+import os
+
+import sqlalchemy
+from alembic import command, script
+from alembic.config import Config
+from alembic.operations.ops import MigrationScript
+from alembic.runtime import migration
+
+import api.logging
+
+logger = api.logging.get_logger(__name__)
+alembic_cfg = Config(os.path.join(os.path.dirname(__file__), "./alembic.ini"))
+
+# Override the script_location to be absolute based on this file's directory.
+alembic_cfg.set_main_option("script_location", os.path.dirname(__file__))
+
+
+def up(revision="head"):
+    command.upgrade(alembic_cfg, revision)
+
+
+def down(revision="-1"):
+    command.downgrade(alembic_cfg, revision)
+
+
+def downall(revision="base"):
+    command.downgrade(alembic_cfg, revision)
+
+
+def have_all_migrations_run(db_engine: sqlalchemy.engine.Engine) -> None:
+    directory = script.ScriptDirectory.from_config(alembic_cfg)
+    with db_engine.begin() as connection:
+        context = migration.MigrationContext.configure(connection)
+        current_heads = set(context.get_current_heads())
+        expected_heads = set(directory.get_heads())
+
+        # Only throw _if_ it's been migrated and doesn't match expectations.
+        # Otherwise, don't bother with this - most likely running in a testing environment.
+        if current_heads != expected_heads:
+            raise Exception(
+                "The database schema is not in sync with the migrations. Please verify that the "
+                "migrations have been run up to {}; currently at {}".format(
+                    expected_heads, current_heads
+                )
+            )
+
+        logger.info(
+            "The current migration head is up to date, {} and Alembic is expecting {}".format(
+                current_heads, expected_heads
+            )
+        )
```

Comment:
I think for logging the generally preferred way is using their formatting (eg. `logger.info("example text %s %s", param1, param2)`) as it does just-in-time formatting and won't bother formatting the string if the log gets thrown away because you have the log level set above it.

A very minor optimization: https://docs.python.org/3/howto/logging.html#optimization - and I think that's what Mass had trended towards. That said, I can't imagine this matters much and f-strings tend to be the easiest to read so I'll switch to that.

Reply:
Neat! I don't have any concerns with using the logger formatting (e.g. `logger.info("example text %s %s", param1, param2)`), but that's not what we have here (e.g. `logger.info("{} string {}".format(param1, param2))`). If we're going to not do the formatting directly using `logger.info` (etc), then yah I think f-strings is easier to read.